// json-ts/dist/index.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var ts = require("typescript");
var parser_1 = require("./parser");
exports.parse = parser_1.parse;
var printer_1 = require("./printer");
exports.print = printer_1.print;
var transformer_1 = require("./transformer");
exports.transform = transformer_1.transform;
var collapse_interfaces_1 = require("./collapse-interfaces");
exports.defaults = {
    prefix: "I",
    rootName: "RootObject"
};
function json2ts(validJsonString, options) {
    if (options === void 0) { options = {}; }
    var mergedOptions = __assign({}, exports.defaults, options);
    var _a = parser_1.parse(validJsonString, mergedOptions), stack = _a.stack, inputKind = _a.inputKind;
    switch (inputKind) {
        case ts.SyntaxKind.ArrayLiteralExpression:
        case ts.SyntaxKind.ObjectLiteralExpression: {
            var transformed = transformer_1.transform(stack, mergedOptions);
            var flattened = collapse_interfaces_1.collapseInterfaces(transformed);
            var printed = printer_1.print(flattened, inputKind, mergedOptions);
            return printed;
        }
        default: {
            var printed = printer_1.printLiteral(stack[0], inputKind, mergedOptions);
            return printed;
        }
    }
}
exports.json2ts = json2ts;
function json2tsMulti(validJsonStrings, options) {
    if (options === void 0) { options = {}; }
    var inputKinds = new Set([]);
    var mergedOptions = __assign({}, exports.defaults, options);
    var joined = validJsonStrings.reduce(function (all, json) {
        var _a = parser_1.parse(json, mergedOptions), stack = _a.stack, inputKind = _a.inputKind;
        inputKinds.add(inputKind);
        var transformed = transformer_1.transform(stack, mergedOptions);
        return all.concat(transformed);
    }, []);
    if (inputKinds.size > 1) {
        // todo handle mixed types
    }
    var flattened = collapse_interfaces_1.collapseInterfaces(joined);
    var printed = printer_1.print(flattened, Array.from(inputKinds)[0], mergedOptions);
    return printed;
}
exports.json2tsMulti = json2tsMulti;
if ((typeof window !== 'undefined') && ((typeof window.json2ts) === 'undefined')) {
    window.json2ts = json2ts;
    window.json2ts.parse = parser_1.parse;
    window.json2ts.transform = transformer_1.transform;
    window.json2ts.print = printer_1.print;
}
