// typedoc-plugin-markdown/dist/theme/theme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs-extra");
const typedoc_1 = require("typedoc");
const reflections_1 = require("typedoc/dist/lib/models/reflections");
const UrlMapping_1 = require("typedoc/dist/lib/output/models/UrlMapping");
const theme_1 = require("typedoc/dist/lib/output/theme");
const DefaultTheme_1 = require("typedoc/dist/lib/output/themes/DefaultTheme");
class MarkdownTheme extends theme_1.Theme {
    constructor(renderer, basePath, options) {
        super(renderer, basePath);
        this.navigationTitlesMap = {};
        this.hasGlobalsFile = false;
        renderer.removeComponent('navigation');
        renderer.removeComponent('assets');
        renderer.removeComponent('javascript-index');
        renderer.removeComponent('toc');
        renderer.removeComponent('pretty-print');
    }
    isOutputDirectory(outputDirectory) {
        let isOutputDirectory = true;
        const allowedListings = [
            this.indexName,
            'globals.md',
            'SUMMARY.md',
            'classes',
            'enums',
            'interfaces',
            'media',
            'modules',
            '.DS_Store',
        ];
        const listings = fs.readdirSync(outputDirectory);
        if (!listings.includes(this.indexName)) {
            isOutputDirectory = false;
            return;
        }
        listings.forEach(listing => {
            if (!allowedListings.includes(listing)) {
                isOutputDirectory = false;
                return;
            }
        });
        return isOutputDirectory;
    }
    getUrls(project) {
        const urls = [];
        const entryPoint = this.getEntryPoint(project);
        if (project.readme && this.application.options.getValue('readme') !== 'none') {
            entryPoint.url = this.globalsName;
            urls.push(new UrlMapping_1.UrlMapping(this.globalsName, entryPoint, 'reflection.hbs'));
            urls.push(new UrlMapping_1.UrlMapping(this.indexName, entryPoint, 'index.hbs'));
            this.hasGlobalsFile = true;
        }
        else {
            entryPoint.url = this.indexName;
            urls.push(new UrlMapping_1.UrlMapping(this.indexName, entryPoint, 'reflection.hbs'));
        }
        if (entryPoint.children) {
            entryPoint.children.forEach((child) => {
                if (child instanceof reflections_1.DeclarationReflection) {
                    this.buildUrls(child, urls);
                }
            });
        }
        this.navigation = this.getNavigation(project);
        if (this.navigation.children) {
            this.navigation.children.forEach(navItem => {
                navItem.children.forEach(navItemChild => {
                    this.navigationTitlesMap[navItemChild.url] = navItemChild.title;
                });
            });
        }
        return urls;
    }
    getNavigation(project) {
        function createNavigationGroup(name, url = null) {
            const navigationGroup = new typedoc_1.NavigationItem(name, url);
            navigationGroup.children = [];
            delete navigationGroup.cssClasses;
            delete navigationGroup.reflection;
            return navigationGroup;
        }
        function getNavigationGroup(reflection) {
            if (reflection.kind === reflections_1.ReflectionKind.ExternalModule) {
                return externalModulesNavigation;
            }
            if (reflection.kind === reflections_1.ReflectionKind.Module) {
                return modulesNavigation;
            }
            if (reflection.kind === reflections_1.ReflectionKind.Class) {
                return classesNavigation;
            }
            if (reflection.kind === reflections_1.ReflectionKind.Enum) {
                return enumsNavigation;
            }
            if (reflection.kind === reflections_1.ReflectionKind.Interface) {
                return interfacesNavigation;
            }
            return null;
        }
        function addNavigationItem(reflection, parentNavigationItem, group) {
            let navigationGroup;
            if (group) {
                navigationGroup = group;
            }
            else {
                navigationGroup = getNavigationGroup(reflection);
            }
            let titlePrefix = '';
            if (parentNavigationItem && parentNavigationItem.title) {
                titlePrefix = parentNavigationItem.title.replace(/\"/g, '') + '.';
            }
            const title = titlePrefix + reflection.name.replace(/\"/g, '');
            const url = reflection.url;
            const nav = new typedoc_1.NavigationItem(title, url, parentNavigationItem);
            nav.parent = parentNavigationItem;
            navigationGroup.children.push(nav);
            if (reflection.children) {
                reflection.children.forEach(reflectionChild => {
                    if (reflectionChild.hasOwnDocument) {
                        addNavigationItem(reflectionChild, nav, navigationGroup);
                    }
                });
            }
            delete nav.cssClasses;
            delete nav.reflection;
            return nav;
        }
        const isModules = this.application.options.getValue('mode') === 1;
        const navigation = createNavigationGroup(project.name, this.indexName);
        const externalModulesNavigation = createNavigationGroup('External Modules');
        const modulesNavigation = createNavigationGroup('Modules');
        const classesNavigation = createNavigationGroup('Classes');
        const enumsNavigation = createNavigationGroup('Enums');
        const interfacesNavigation = createNavigationGroup('Interfaces');
        if (!isModules) {
            project.groups.forEach(group => {
                group.children.forEach(reflection => {
                    if (reflection.hasOwnDocument) {
                        addNavigationItem(reflection);
                    }
                });
            });
        }
        if (isModules) {
            project.groups[0].children.forEach(module => {
                const moduleNavigation = addNavigationItem(module);
                if (module.children) {
                    module.children.forEach(reflection => {
                        if (reflection.hasOwnDocument) {
                            addNavigationItem(reflection, moduleNavigation);
                        }
                    });
                }
            });
        }
        if (externalModulesNavigation.children.length) {
            navigation.children.push(externalModulesNavigation);
        }
        if (modulesNavigation.children.length) {
            navigation.children.push(modulesNavigation);
        }
        if (classesNavigation.children.length) {
            navigation.children.push(classesNavigation);
        }
        if (enumsNavigation.children.length) {
            navigation.children.push(enumsNavigation);
        }
        if (interfacesNavigation.children.length) {
            navigation.children.push(interfacesNavigation);
        }
        return navigation;
    }
    getEntryPoint(project) {
        const entryPoint = this.owner.entryPoint;
        if (entryPoint) {
            const reflection = project.getChildByName(entryPoint);
            if (reflection) {
                if (reflection instanceof reflections_1.ContainerReflection) {
                    return reflection;
                }
                else {
                    this.application.logger.warn('The given entry point `%s` is not a container.', entryPoint);
                }
            }
            else {
                this.application.logger.warn('The entry point `%s` could not be found.', entryPoint);
            }
        }
        return project;
    }
    buildUrls(reflection, urls) {
        const mapping = DefaultTheme_1.DefaultTheme.getMapping(reflection);
        if (mapping) {
            if (!reflection.url || !DefaultTheme_1.DefaultTheme.URL_PREFIX.test(reflection.url)) {
                const url = `${mapping.directory}/${DefaultTheme_1.DefaultTheme.getUrl(reflection)}.md`;
                urls.push(new UrlMapping_1.UrlMapping(url, reflection, mapping.template));
                reflection.url = url;
                reflection.hasOwnDocument = true;
            }
            for (const child of reflection.children || []) {
                if (mapping.isLeaf) {
                    this.applyAnchorUrl(child, reflection);
                }
                else {
                    this.buildUrls(child, urls);
                }
            }
        }
        else if (reflection.parent) {
            this.applyAnchorUrl(reflection, reflection.parent);
        }
        return urls;
    }
    applyAnchorUrl(reflection, container) {
        if (!reflection.url || !DefaultTheme_1.DefaultTheme.URL_PREFIX.test(reflection.url)) {
            reflection.url = container.url + '#' + this.getAnchor(reflection);
            reflection.hasOwnDocument = false;
        }
        reflection.traverse(child => {
            if (child instanceof reflections_1.DeclarationReflection) {
                this.applyAnchorUrl(child, container);
            }
        });
    }
    getAnchor(reflection) {
        return MarkdownTheme.getAnchorRef(reflection);
    }
    static getAnchorRef(reflection) {
        function parseAnchorRef(ref) {
            return ref.replace(/"/g, '').replace(/ /g, '-');
        }
        let anchorPrefix = '';
        reflection.flags.forEach(flag => (anchorPrefix += `${flag}-`));
        const prefixRef = parseAnchorRef(anchorPrefix);
        const reflectionRef = parseAnchorRef(reflection.name);
        const anchorRef = prefixRef + reflectionRef;
        return anchorRef.toLowerCase();
    }
    get indexName() {
        return 'README.md';
    }
    get globalsName() {
        return 'globals.md';
    }
}
exports.MarkdownTheme = MarkdownTheme;
